crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -755,147 +769,186 @@ where
     let segment_headers = Arc::clone(&subspace_link.segment_headers);
+    let mut archiver_notification_stream = subspace_link.archiver_notification_stream.subscribe();
 
-    Ok(async move {
-        // Farmers may have not received all previous segments, send them now.
-        for archived_segment in older_archived_segments {
-            send_archived_segment_notification(
-                &archived_segment_notification_sender,
-                archived_segment,
-            )
-            .await;
-        }
+    Ok({
+        let subspace_link = subspace_link.clone();
+        async move {
+            // Farmers may have not received all previous segments, send them now.
+            for archived_segment in older_archived_segments {
+                send_archived_segment_notification(
+                    &archived_segment_notification_sender,
+                    archived_segment,
+                )
+                .await;
+            }
 
-        while let Some(BlockImportingNotification {
-            block_number,
-            // Just to be very explicit that block import shouldn't continue until archiving
-            // is over
-            acknowledgement_sender: _acknowledgement_sender,
-            ..
-        }) = block_importing_notification_stream.next().await
-        {
-            let block_number_to_archive =
-                match block_number.checked_sub(&confirmation_depth_k.into()) {
-                    Some(block_number_to_archive) => block_number_to_archive,
-                    None => {
-                        continue;
+            while let Some(BlockImportingNotification {
+                block_number,
+                // Just to be very explicit that block import shouldn't continue until archiving
+                // is over
+                acknowledgement_sender: _acknowledgement_sender,
+            }) = block_importing_notification_stream.next().await
+            {
+                match archiver_notification_stream.try_recv() {
+                    Ok(data) => {
+                        debug!(?data, "Archiver notification event received.");
+                        let last_archived_block = data.last_archived_block;
+
+                        let last_archived_segment = last_archived_block.0;
+                        best_archived_block_number =
+                            last_archived_block.0.last_archived_block().number.into();
+                        best_archived_block_hash = last_archived_block.1.block.header().hash();
+
+                        let last_archived_block_encoded = encode_block(last_archived_block.1);
+                        archiver = Archiver::with_initial_state(
+                        archiver.kzg(),
+                        last_archived_segment,
+                        &last_archived_block_encoded,
+                        last_archived_block.2,
+                    ).expect("Invalid initial archival state should stop the application to prevent further losses.");
```

Comment:
rustfmt is not able to format code if you have too long strings. Split it into more than one line and it'll be able to successfully reformat the code. It looks a bit off right now.